core/src/main/java/org/apache/iceberg/expressions/ExpressionParser.java

```diff
@@ -257,6 +305,28 @@ private void transform(String transform, String name) throws IOException {
       gen.writeStringField(TERM, name);
       gen.writeEndObject();
     }
+
+    private void transformWithFieldId(String transform, String name, int fieldId)
```

Comment:
It seems strange to duplicate most of `transform` here. Why not add a `Integer` parameter and write the field ID if it is non-null? Then this would just need to pass both `name` and `fieldId` to a single `reference` method.